src/misc/umap.c

```diff
@@ -0,0 +1,864 @@
+ * derivative/force. Same applies for the repulsion */
+static igraph_error_t igraph_i_umap_attract(igraph_real_t xd, igraph_real_t yd, igraph_real_t mu,
+       igraph_real_t a, igraph_real_t b, igraph_real_t *force_x, igraph_real_t *force_y)
+{
+    igraph_real_t dsq, phi, force;
+
+    dsq = xd * xd + yd * yd;
+
+    phi = 1. / (1. + a * pow(dsq, b));
+    force = - mu * (2 * a * b * pow(dsq, b - 1)) * phi;
+    *force_x = force * xd;
+    *force_y = force * yd;
+
+    /* clip force to avoid too rapid change */
+    igraph_i_umap_clip_force(force_x, 3);
+    igraph_i_umap_clip_force(force_y, 3);
+
+#ifdef UMAP_DEBUG
+    printf("force attractive: xd = %f, fx = %f\n", xd, *force_x);
+    printf("force attractive: yd = %f, fy = %f\n", yd, *force_y);
+#endif
+
+    return IGRAPH_SUCCESS;
+}
+
+/*xd is difference in x direction, mu is a weight */
+static igraph_error_t igraph_i_umap_repel(igraph_real_t xd, igraph_real_t yd, igraph_real_t mu,
+       igraph_real_t a, igraph_real_t b, igraph_real_t *force_x, igraph_real_t *force_y)
+{
+    igraph_real_t dsq, force;
+    igraph_real_t min_dist = 0.01;
+
+    dsq = fmax(min_dist * min_dist, xd * xd + yd * yd);
+
+    /* NOTE: in practice, in negative sampling mu is always zero because we
+     * *assume* the sample to be negative i.e. never a true edge */
+    force = (1 - mu) * (2 * b) / dsq / (1 + a * pow(dsq, b));
+    *force_x = force * xd;
+    *force_y = force * yd;
+
+    /* clip force to avoid too rapid change */
+    igraph_i_umap_clip_force(force_x, 3);
+    igraph_i_umap_clip_force(force_y, 3);
+
+#ifdef UMAP_DEBUG
+    printf("force repulsive: xd = %f, fx = %f\n", xd, *force_x);
+    printf("force repulsive: yd = %f, fy = %f\n", yd, *force_y);
+#endif
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t igraph_i_umap_apply_forces(const igraph_t *graph,  const igraph_vector_t *umap_weights,
+       igraph_matrix_t *layout, igraph_real_t a, igraph_real_t b, igraph_real_t prob,
+       igraph_real_t learning_rate, igraph_bool_t avoid_neighbor_repulsion)
+{
+    igraph_integer_t no_of_nodes = igraph_matrix_nrow(layout);
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_integer_t from, to;
+    igraph_real_t from_x, from_y, to_x, to_y, x_diff, y_diff, fx, fy;
+    /* The following is only used for small graphs, to avoid repelling your neighbors
+     * For large sparse graphs, it's not necessary. For large dense graphs, you should
+     * not be doing UMAP.
+     * */
+    igraph_vector_int_t neis;
+
+    if (avoid_neighbor_repulsion) {
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&neis, 0);
+    }
+
+    igraph_integer_t n_random_vertices = sqrt(no_of_nodes);
+
+    /* iterate over a random subsample of edges */
+    for (igraph_integer_t eid = 0; eid < no_of_edges; eid++) {
+        if (RNG_UNIF01() > prob) {
+            continue;
+        }
+
+        /* half the time, swap the from/to, otherwise some vertices are never moved */
```

Comment:
Why is only one endpoint of the edge considered? I see that the paper says the same,

> Thus, whenever an attractive force is applied to an edge, one of that edge’s vertices is repulsed by a sampling of other vertices

but I'm still confused.

Reply:
I'm confused as well, and this might be a source of numerical instabilities. I'm investigating...

Reply:
Seems like the algo is now actually stable enough to converge on all platforms. Part of it was clipping the gradient to avoid giant swings.

As for this particular observation, I think I understand what's going on. At every iteration (epoch), only some edges are adjusted to save time. Indeed, the whole algorithm does not "converge" per se, it's more like we reduce the learning rate/temperature until it becomes zero and at that point the layout is frozen in place. Whether that can be improved is a good question.... I guess so? At the same time, UMAP has rotational invariance, so perhaps if we don't freeze it it might start spinning rigidly? In any case, even if an edge is not picked for shift in this epoch, it might be picked in the next one, and the temperature is almost identical so it's alright. Same goes for the vertices that compose the edge: even if one is not picked this epoch, it might be in the next epoch and the difference is almost nothing. Of course, we could always update the vertices of an edge at the same time, but then both of them would be repelled by the rest of the graph, which is computationally as expensive as just waiting out another epoch and doing the other vertex. In addition, by moving both vertices simultaneously we add one more constraint to the minimization, which might somewhat reduce the ability of the algo to explore the solution space efficiently.

In conclusion, I think it's arbitrary but so is the choice of the edge sampling probability, and increasing the latter is almost equivalent to changing the former (i.e. moving both vertices of the edge). So it's okay for now I'd say.

If we do proper profiling in the future, we might be able to cut corners for speed and this might be one of them.